SNP filtering - for population structure
================

Remove indels:

``` bash
cd ./turqueti/tarcap/SNPfiltering1/1_rmindel

/sw/containers/vcftools-0.1.16.sif vcftools --gzvcf ./turqueti/tarcap/bcftools_mpileup/raw.vcf.gz --remove-indels --out rmindels --recode --recode-INFO-all
```

Remove technical duplicates (check for potential batch effect using PCA
first):

``` bash
cd ./turqueti/tarcap/SNPfiltering1/2_rm_tech_dups

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/1_rmindel/rmindels.recode.vcf --remove rm_tech_dups.tsv --out rmduplciates.vcf --recode --recode-INFO-all

## Calculate missing data per indiviudal
/sw/containers/vcftools-0.1.16.sif vcftools --vcf rmduplciates.vcf.recode.vcf --missing-indv --out out

## Calculate mean depth per site
/sw/containers/vcftools-0.1.16.sif vcftools --vcf rmduplciates.vcf.recode.vcf --site-mean-depth --out out

awk '{ total += $3 } END { print total/NR }' out.ldepth.mean #mean depth = 48.1826
```

Remove samples with high missing data (>80%):

``` bash
# Skip, we don't have samples with high missing data. 
```

Filter by minimum base quality score of 30:

``` bash
cd ./turqueti/tarcap/SNPfiltering1/4_minQ

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/4_minQ/2_rm_tech_dups/rmduplciates.vcf.recode.vcf --minQ 30 --out minQ30 --recode --recode-INFO-all
```

Remove sites with mean depth less than average/3 (‐‐min-meanDP 16) and
greater than average\*2 (–max-meanDP 96):

``` bash
cd ./turqueti/tarcap/SNPfiltering1/5_site_depth

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/4_minQ/minQ30.recode.vcf --min-meanDP 16 --max-meanDP 96 --out site_depth --recode --recode-INFO-all
```

Keep biallelic sites only:

``` bash
cd ./turqueti/tarcap/SNPfiltering1/6_biallelic_sites

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/5_site_depth/site_depth.recode.vcf --min-alleles 2 --max-alleles 2 --out biallelic_sites --recode --recode-INFO-all
```

Keep sites present in 50% of all samples (i.e. filter by 50% missing
data):

``` bash
cd ./turqueti/tarcap/SNPfiltering1/7_max_missing05

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/6_biallelic_sites/biallelic_sites.recode.vcf --max-missing 0.5 --out maxmissing05 --recode --recode-INFO-all
```

Keep sites with minor allele frequency at least 0.05:

``` bash
cd ./turqueti/tarcap/SNPfiltering1/8_maf005

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/7_max_missing05/maxmissing05.recode.vcf --maf 0.05 --out maf005 --recode --recode-INFO-all
```

Keep sites with a maximum observed heterozygosity value of 0.5: Typical
for RAD-loci data, for refs see
<https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x>
& <https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13314>

``` bash
## first, define site ID
/sw/containers/bcftools-1.13.sif bcftools annotate --set-id '%CHROM\_%POS' maf005.recode.vcf > maf005_setID.vcf
```

``` r
## calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

### get a genind from the semi-filtered vcf file from bcftools

vcf2 <- read.vcfR("./maf005_setID.vcf")
tur_genind <- vcfR2genind(vcf2)
summary_genind = summary(tur_genind)

### get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

### you can then sort the values in excel and save a list of the SNP (with site ID) with >0.5
```

``` bash
## Filter sites with maximum observed heterozygosity of 0.5
cd ./turqueti/tarcap/SNPfiltering1/9_max_obs_het05

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/8_maf005/maf005_setID.vcf --exclude site_obs_het_over_05.tsv --out max_het05.vcf --recode --recode-INFO-all
```

Get 1 SNP per locus: In here, 1000 is the arbitrary value larger than
the longest contig in unique_consensus_loci.fa

``` bash
cd ./turqueti/tarcap/SNPfiltering1/10_thin1000

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/9_max_obs_het05/max_het05.vcf.recode.vcf --thin 1000 --out thin1000_maf005_obshet05.vcf --recode --recode-INFO-all
```

## SNP filtering - for Admixtools

start from maxmissing05.recode.vcf in
./turqueti/tarcap/SNPfiltering1/7_max_missing05

``` bash
cd ./turqueti/tarcap/SNPfiltering1/admixtools/8_mac001

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/7_max_missing05/maxmissing05.recode.vcf --mac 1 --out mac1 --recode --recode-INFO-all
```

keep sites with a maximum observed heterozygosity value of 0.7: typical
for RAD-loci data, for refs see
<https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x>
& <https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13314>

``` bash
## first, define site ID
/sw/containers/bcftools-1.13.sif bcftools annotate --set-id '%CHROM\_%POS' mac1.recode.vcf > mac1_setID.vcf
```

``` r
## calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

### get a genind from the semi-filtered vcf file from bcftools

vcf2 <- read.vcfR("./mac1_setID.vcf")
tur_genind <- vcfR2genind(vcf2)
summary_genind = summary(tur_genind)

### get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

### you can then sort the values in excel and save a list of the SNP (with site ID) with >0.5
```

``` bash
# filter sites with maximum observed heterozygosity of 0.7
cd ./turqueti/tarcap/SNPfiltering1/admixtools/9_max_obs_het07

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/admixtools/8_mac001/mac1_setID.vcf --exclude site_obs_het_over_07.tsv --out max_het07.vcf --recode --recode-INFO-all
```

keep admixtools samples

``` bash
/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/admixtools/9_max_obs_het07/max_het07.vcf.recode.vcf --keep keep_admixtools_samples.tsv --out admixtools.vcf --recode --recode-INFO-all
```

## SNP filtering - for GDM (all P. turqueti samples)

start from max_het05.vcf.recode.vcf in cd
./turqueti/tarcap/SNPfiltering1/9_max_obs_het05

``` bash
#remove outgroup species
cd ./turqueti/tarcap/SNPfiltering1/GDM/10_rmoutgroup

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/9_max_obs_het05/max_het05.vcf.recode.vcf --remove rm_outgroup.tsv --out GDM_tur --recode --recode-INFO-all

module load htslib
bgzip GDM_tur.recode.vcf
tabix GDM_tur.recode.vcf.gz

bcftools view -f PASS --types snps -m1 -M2 -r `seq -s, 1 22` GDM_tur.recode.vcf.gz | singularity run $SING/plink-1.90b6.26.sif plink1.9 --double-id --vcf GDM_tur.recode.vcf.gz --recode A-transpose --allow-extra-chr --out ./ngsDist/GDM_tur

tail -n +2 GDM_tur.traw | cut -f 1,4- | perl -p -e 's/\bNA\b/-1/g' | gzip > ngsDist.geno.gz

zgrep CHROM ./GDM_tur.recode.vcf.gz | cut -f 10- | tr "\t" "\n" > ngsDist.labels

#calculate pairwise genetic distance using nsgDist 

./ngsDist --geno ./ngsDist.geno.gz --n_ind 96 --n_sites 37322 --labels ./ngsDist.labels --out ./GDM_tur.dist 
```

## SNP filtering - for GDM (for P. turqueti samples from the Scotia Sea only)

start from max_het05.vcf.recode.vcf in cd
./turqueti/tarcap/SNPfiltering1/9_max_obs_het05

``` bash
#remove outgroup species + non Scotia Sea samples 
cd ./turqueti/tarcap/SNPfiltering1/GDM/10_rmoutgroup_nonScotia

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/9_max_obs_het05/max_het05.vcf.recode.vcf --remove rm_outgroup_nonScotia.tsv --out GDM_scotia --recode --recode-INFO-all

module load htslib
bgzip GDM_scotia.recode.vcf
tabix GDM_scotia.recode.vcf.gz

bcftools view -f PASS --types snps -m1 -M2 -r `seq -s, 1 22` GDM_scotia.recode.vcf.gz | singularity run $SING/plink-1.90b6.26.sif plink1.9 --double-id --vcf GDM_scotia.recode.vcf.gz --recode A-transpose --allow-extra-chr --out ./ngsDist/GDM_scotia

tail -n +2 GDM_scotia.traw | cut -f 1,4- | perl -p -e 's/\bNA\b/-1/g' | gzip > ngsDist_scotia.geno.gz

zgrep CHROM ./GDM_scotia.recode.vcf.gz | cut -f 10- | tr "\t" "\n" > ngsDist_scotia.labels

#calculate pairwise genetic distance using nsgDist 

./ngsDist --geno ./ngsDist_scotia.geno.gz --n_ind 52 --n_sites 37322 --labels ./ngsDist_scotia.labels --out ./GDM_scotia.dist 
```
