---
title: "SNP filtering - demographic models / stairwayplot"
output: github_document
---
Keep samples from Weddell Sea, Amundsen Sea, Ross Sea amd East Antarctica (4 population model in main text):
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/SNPfiltering2/1_keep_seaway_samples

bcftools view -S keep_4popseaway_samples.txt  -o seawayonly-vcf.vcf ./turqueti/tarcap/bcftools_mpileup/raw.vcf.gz
```

Remove sites with mean depth less than 5 (‐‐min-meanDP 5) and greater than average*2 (--max-meanDP 103):
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/SNPfiltering2/2_site_depth

vcftools --vcf ./1_keep_seaway_samples/seawayonly-vcf.vcf --min-meanDP 5 --max-meanDP 103 --out site_depth --recode --recode-INFO-all

```

Remove indels:
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/SNPfiltering2/3_rm_indels

vcftools --vcf ./2_site_depth/site_depth.recode.vcf --remove-indels --out rmindels --recode --recode-INFO-all

#keep polymorphic sites only
bcftools view -c 1 rmindels.recode.vcf -Ov -o rmindels_polyonly.vcf

#keep monomorphic sites only (for number of mono site counting purpose)
bcftools view -C 0 rmindels.recode.vcf -Ov -o rmindels_monoonly.vcf
```

From here on, only filter polymorphic sites
Filter sites with minQ20
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/SNPfiltering2/4_minQ20

vcftools --vcf ./3_remove_indels/rmindels_polyonly.vcf --minQ 20 --out minQ20polyonly --recode --recode-INFO-all
```

Filter sites with --hwe 0.0001:
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/SNPfiltering2/5_hwe

vcftools --vcf ./4_minQ20/minQ20polyonly.recode.vcf --hwe 0.0001 --out hewpolyonly --recode --recode-INFO-all
```

Keep biallelic SNPs only:
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/SNPfiltering2/6_biallelic

vcftools --vcf ./5_hwe/hewpolyonly.recode.vcf --max-alleles 2 --out biallelicpolyonly --recode --recode-INFO-all
#no site was lost
```

Keep sites with a maximum observed heterozygosity value of 0.7:
```{bash, error=TRUE, results='hide'}
## first, define site ID
bcftools annotate --set-id '%CHROM\_%POS' biallelicpolyonly.recode.vcf > biallelicpolyonly_setID.vcf

```
```{r, eval=F, echo=T}
## calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

### get a genind from the semi-filtered vcf file from bcftools

vcf2 <- read.vcfR("./biallelicpolyonly_setID.vcf")
tur_genind <- vcfR2genind(vcf2)
summary_genind = summary(tur_genind)

### get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

### sort the values in excel and save a list of the SNP (with site ID) with >0.7
```
```{bash, error=TRUE, results='hide'}
## Filter sites with maximum observed heterozygosity of 0.7
cd ./turqueti/tarcap/SNPfiltering2/7_max_obs_het07

vcftools --vcf ./6_biallelic/biallelicpolyonly_setID.vcf --exclude site_obs_het_over_07.tsv --out max_het07_polyonly.vcf --recode --recode-INFO-all
```

Set ancestral alelles using outgroups in vcf file:  
See ./hpc/tarcap/set_ancestral_alleles.Rmd for further info
```{bash, error=TRUE, results='hide'}
## create an INFO file line for the new vcf file
cd ./turqueti/tarcap/SNPfiltering2/9_setAA

echo '##INFO=<ID=AA,Number=1,Type=Character,Description="Ancestral allele">' > hdr.txt

## use bcftools to annotate the vcf file with the ancestral allele information
bcftools annotate -a ./out_aa.tab.gz \
 -c CHROM,POS,REF,ALT,INFO/AA -h hdr.txt -Ov \
 -o max_het07_polyonly_aa.vcf max_het07_polyonly.vcf.recode.vcf
```