Demographic modelling with Fastsimcoal 3 populations
================

## Prepare input files for fastsimcoal

``` bash
cd ./turqueti/tarcap/analyses/fastsimcoal/vcf2sfs_SHEWSRS

python sampleKgenotypesPerPop.py -i ./SHEWSRS_max_het07_polyonly_aa.vcf -o seaway_tur_SHEWSRS_resampled.vcf -p SHEWSRS_original_popmap.txt -k Ross:8,SHE:13,Weddell:15 -v
```

Generate multiSFS for fastsimcoal  
<br>

``` bash
python ./vcf2sfs.py -i ./seaway_tur_SHEWSRS_resampled.vcf -o SHE_WS_RS -p ./seaway_tur_SHEWSRS_resampled_popmap.txt -q SHE,Weddell,Ross -v

python ./vcf2sfs.py -i ./seaway_tur_SHEWSRS_resampled.vcf -o WS_RS_SHE -p ./seaway_tur_SHEWSRS_resampled_popmap.txt -q Weddell,Ross,SHE -v
```

## Specify fastsimcoal models

<br> The models that were tested in the paper are illustrated in the
supplementary figures.  
<br> See hpc/fastsimcol_3pops_SHE/models in this repo for corresponding
.est and .tpl files  
<br>

## Run fastsimcoal models

Run fastsimocal. Do 100 reps with 200,000 coalescent simulations per
model e.g.Â SHE_WSRS

``` bash
PREFIX="EA_WSRS"

cd ./turqueti/tarcap/analyses/fastsimcoal/models/SHE_WSRS #where the .est, .tpl and .obs are

 for i in {1..100}
 do
   mkdir run$i
   cp ${PREFIX}.tpl ${PREFIX}.est *.obs run$i"/"
   cd run$i
   ./fsc27093 -t ${PREFIX}.tpl -e ${PREFIX}.est -M -n200000 -l50 -L50 -C10 -c10 -B10 -d --multiSFS &> log.oe
   cd ..
 done
```

Find the best run using fsc-selectbestrun.sh  
fsc-selectbestrun.sh is from
<https://speciationgenomics.github.io/fastsimcoal2/>

``` bash
cd ./turqueti/tarcap/analyses/fastsimcoal/models/SHE_WSRS

bash fsc-selectbestrun.sh
```
